### PR TITLE
Service principal login option

### PR DESCRIPTION
**Pull request by zhang-hua**
# Type of change? 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

...

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -3,17 +3,166 @@
...
-        "sourceMaps": true,
-        "program": "${workspaceRoot}/src/index.ts",
-        //"args": [ "-r", "java", "-s", "standard" ],
-        "args": [ "-t", "test", "-s", "test" ],
-        "console": "integratedTerminal",
-        "outFiles": [
-            "${workspaceRoot}/publish/**/*.js"
-        ]
-    }]
+    "configurations": [
```
Comment (ppathan) [inline .vscode/launch.json]: Add this file to .gitignore after merge

File: src/index.ts [hunk 2]
```diff
@@ -12,8 +12,13 @@ import * as momemt from 'moment';
 
 import { exec } from 'child_process';
 import { ChoiceType, prompt } from 'inquirer';
-import { AuthResponse, AzureEnvironment, DeviceTokenCredentials, DeviceTokenCredentialsOptions,
-    LinkedSubscription, InteractiveLoginOptions, interactiveLoginWithAuthResponse } from 'ms-rest-azure';
+import {
+    AuthResponse, AzureEnvironment, AzureTokenCredentialsOptions, DeviceTokenCredentials, DeviceTokenCredentialsOptions,
```
Comment (ppathan) [inline src/index.ts]: nit: put them on single line

File: src/index.ts [hunk 3]
```diff
@@ -182,30 +224,63 @@ function main() {
...
+                        serviceClientCredentials,
+                        cachedAuthResponse.credentials.environment,
+                        answers.subscriptionId,
+                        program.type,
+                        program.sku);
                     return deploymentManager.getLocations();
                 })
                 .then((locations: string[] | undefined) => {
-                    if (locations && locations.length > 0) {
+                    if (program.location && (program.websiteName || program.solutionName) && program.username) {
```
Comment (ppathan) [inline src/index.ts]: if one of program.websiteName or program.solutionName is null we should set them to the non-null value and use below for azureWebsiteName and solutionName

File: src/index.ts [hunk 4]
```diff
@@ -182,30 +224,63 @@ function main() {
...
+                    } else if (locations && locations.length > 0) {
                         return prompt(getDeploymentQuestions(locations));
                     }
                     throw new Error('Locations list cannot be empty');
                 })
                 .then((ans: Answers) => {
-                    answers.location = ans.location;
-                    answers.azureWebsiteName = ans.azureWebsiteName;
-                    answers.adminUsername = ans.adminUsername;
+                    answers = {...answers, ...ans};
```
Comment (ppathan) [inline src/index.ts]: Sweet!

File: src/index.ts [hunk 5]
```diff
@@ -329,105 +381,162 @@ function logout() {
...
     let objectId: string = '';
-    return graphClient.applications.create(applicationCreateParameters)
+
+    let servicePrincipalCreateParameters;
+    let graphClient: GraphRbacManagementClient;
+
+    options.tokenAudience = 'graph';
+    let graphClientPromise: Promise<GraphRbacManagementClient>;
+    if (usingServicePrincipal) {
+        graphClientPromise = loginWithServicePrincipalSecretWithAuthResponse(
```
Comment (ppathan) [inline src/index.ts]: Is this needed to get a token with graph audience?

File: src/index.ts [hunk 6]
```diff
@@ -329,105 +381,162 @@ function logout() {
...
+            availableToOtherTenants: false,
+            displayName: azureWebsiteName,
+            homepage,
+            identifierUris,
+            oauth2AllowImplicitFlow: true,
+            replyUrls,
+            requiredResourceAccess
+        };
+
+        applicationCreateParameters.passwordCredentials = [{
```
Comment (ppathan) [inline src/index.ts]: if we are creating passwordCredentials everytime then we can merge it with applicationCreateParameters initialization

File: src/index.ts [hunk 7]
```diff
@@ -329,105 +381,162 @@ function logout() {
...
+            options)
+            .then((response: AuthResponse) => {
+                return new GraphRbacManagementClient(response.credentials, options.domain || '', baseUri);
+            });
+    } else {
+        graphClient = new GraphRbacManagementClient(new DeviceTokenCredentials(options), options.domain || '', baseUri);
+        graphClientPromise = Promise.resolve<GraphRbacManagementClient>(graphClient);
+    }
+
+    return graphClientPromise.then((client: GraphRbacManagementClient) => {
```
Comment (ppathan) [inline src/index.ts]: nit: Put .then on newline

File: src/index.ts [hunk 8]
```diff
@@ -329,105 +381,162 @@ function logout() {
...
     .then((result: any) => {
-        const servicePrincipalCreateParameters = {
+        servicePrincipalCreateParameters = {
             accountEnabled: true,
             appId: result.appId
         };
         objectId = result.objectId;
         return graphClient.servicePrincipals.create(servicePrincipalCreateParameters);
     })
+    .then((result: ServicePrincipal) => {
```
Comment (ppathan) [inline src/index.ts]: Do we still need this then block? If we are creating the servicePrincipal everytime then value returned on line 515 should be same as result, no?

File: .vscode/launch.json (hunk 1 above)
Comment (zhang-hua) [reply .vscode/launch.json]: yep.

File: src/index.ts (hunk 2 above)
Comment (zhang-hua) [reply src/index.ts]: It is too long in a single line and can not pass tslint check.

Comment (ppathan) [reply src/index.ts]: I meant put each import on a new line.
```
AuthResponse,
AzureEnvironment,
...
```

File: src/index.ts (hunk 3 above)
Comment (ppathan) [reply src/index.ts]: @zhang-hua Can you add a todo to fix this later?

**Pull request merged**